app/src/test/java/org/simple/clinic/activity/CloseActivitiesWhenUserIsUnauthorizedTest.kt

```diff
@@ -0,0 +1,234 @@
+package org.simple.clinic.activity
+
+import android.app.Activity
+import com.nhaarman.mockito_kotlin.mock
+import com.nhaarman.mockito_kotlin.never
+import com.nhaarman.mockito_kotlin.reset
+import com.nhaarman.mockito_kotlin.verify
+import com.nhaarman.mockito_kotlin.whenever
+import io.reactivex.subjects.PublishSubject
+import org.junit.Before
+import org.junit.Test
+import org.simple.clinic.user.UserSession
+import org.simple.clinic.util.scheduler.TrampolineSchedulersProvider
+
+class CloseActivitiesWhenUserIsUnauthorizedTest {
+
+  private val userUnauthorizedSubject = PublishSubject.create<Boolean>()
+  private val userSession = mock<UserSession>()
+
+  private val closeActivitiesWhenUserIsUnauthorized = CloseActivitiesWhenUserIsUnauthorized(userSession, TrampolineSchedulersProvider())
+
+  @Before
+  fun setUp() {
+    whenever(userSession.isUserUnauthorized())
+        .thenReturn(userUnauthorizedSubject)
+
+    closeActivitiesWhenUserIsUnauthorized.listen()
+  }
+
+  @Test
+  fun `any created activity that is not TheActivity must be finished if the user gets unauthorized`() {
+    // given
+    val aCreatedTheActivity = mock<TheActivity>()
+    val aCreatedActivity = mock<Activity>()
+    simulateActivityCreated(aCreatedActivity)
+    simulateActivityCreated(aCreatedTheActivity)
+
+    // then
+    userUnauthorizedSubject.onNext(false)
+    verify(aCreatedTheActivity, never()).finish()
+    verify(aCreatedActivity, never()).finish()
+    reset(aCreatedTheActivity, aCreatedActivity)
+
+    userUnauthorizedSubject.onNext(true)
+    verify(aCreatedTheActivity, never()).finish()
+    verify(aCreatedActivity).finish()
+    reset(aCreatedTheActivity, aCreatedActivity)
+
+    // This block is intentionally repeated to ensure that the user session unauthorized stream
+    // is observed continuously and not just for the one emission.
+
+    userUnauthorizedSubject.onNext(true)
+    verify(aCreatedTheActivity, never()).finish()
+    verify(aCreatedActivity).finish()
+  }
+
+  @Test
+  fun `any started activity that is not TheActivity must be finished if the user gets unauthorized`() {
+    // given
+    val aStartedTheActivity = mock<TheActivity>()
+    val aStartedActivity = mock<Activity>()
+
+    simulateActivityStarted(aStartedActivity)
+    simulateActivityStarted(aStartedTheActivity)
+
+    // then
+    userUnauthorizedSubject.onNext(false)
+    verify(aStartedTheActivity, never()).finish()
+    verify(aStartedActivity, never()).finish()
+    reset(aStartedTheActivity, aStartedActivity)
+
+    userUnauthorizedSubject.onNext(true)
```

Comment:
Can we replace the `Boolean` with a `Enum`?

Reply:
This is going to affect a lot of classes since many are using the `isUserUnauthorized()` method. I'll do this as a separate story?

https://www.pivotaltracker.com/story/show/167937511